File: yuzu/lib/Diagnostics/SourceMap.cc
#include "yuzu/Diagnostics/SourceMap.h"

#include "yuzu/Diagnostics/Span.h"

#include <algorithm>
#include <cassert>
#include <cstdint>
#include <string>
#include <string_view>
#include <utility>
#include <vector>

namespace yuzu::diagnostics {

namespace {

/// Build an index of line-start code-point offsets for `text`. The first
/// line always starts at 0; each `\n` introduces a new entry pointing
/// one past the newline. A trailing newline therefore produces an empty
/// trailing line, which matches what users expect when they see "line
/// N" in their editor.
std::vector<uint32_t> indexLines(std::u32string_view text) {
  std::vector<uint32_t> starts;
  starts.push_back(0);
  for (uint32_t i = 0; i < text.size(); ++i) {
    if (text[i] == U'\n') {
      starts.push_back(i + 1);
    }
  }
  return starts;
}

} // namespace

SourceId SourceMap::add(std::string name, std::u32string text) {
  std::vector<uint32_t> starts = indexLines(text);
  entries.push_back(Entry{
      .name = std::move(name),
      .text = std::move(text),
      .lineStarts = std::move(starts),
  });
  // SourceId is 1-indexed so a default-constructed `SourceId{}` (value
  // 0) remains an invalid sentinel callers can't accidentally collide
  // with.
  return SourceId{static_cast<uint32_t>(entries.size())};
}

std::string_view SourceMap::getName(SourceId id) const {
  assert(id.value > 0 && id.value <= entries.size() &&
         "SourceId not issued by this SourceMap");
  return entries[id.value - 1].name;
}

std::u32string_view SourceMap::getText(SourceId id) const {
  assert(id.value > 0 && id.value <= entries.size() &&
         "SourceId not issued by this SourceMap");
  return entries[id.value - 1].text;
}

LineCol SourceMap::getLineCol(SourceId id, uint32_t offset) const {
  assert(id.value > 0 && id.value <= entries.size() &&
         "SourceId not issued by this SourceMap");
  const Entry &entry = entries[id.value - 1];

  // Find the largest lineStarts entry <= offset. `upper_bound` returns
  // the first strictly greater entry; one back is the line containing
  // offset. Always non-empty because `lineStarts` always begins with 0.
  const auto it = std::upper_bound(entry.lineStarts.begin(),
                                   entry.lineStarts.end(), offset);
  const auto lineIndex =
      static_cast<uint32_t>((it - entry.lineStarts.begin()) - 1);

  return LineCol{
      .line = lineIndex + 1,
      .column = (offset - entry.lineStarts[lineIndex]) + 1,
  };
}

std::u32string_view SourceMap::getLineText(SourceId id, uint32_t line) const {
  assert(id.value > 0 && id.value <= entries.size() &&
         "SourceId not issued by this SourceMap");
  const Entry &entry = entries[id.value - 1];

  if (line == 0 || line > entry.lineStarts.size()) {
    return {};
  }

  const uint32_t startOffset = entry.lineStarts[line - 1];
  const uint32_t endOffset = line < entry.lineStarts.size()
                                 ? entry.lineStarts[line]
                                 : static_cast<uint32_t>(entry.text.size());

  // Trim the trailing newline (and an optional `\r` before it) so the
  // renderer doesn't smear the snippet onto two lines.
  uint32_t lineEnd = endOffset;
  if (lineEnd > startOffset && entry.text[lineEnd - 1] == U'\n') {
    --lineEnd;
    if (lineEnd > startOffset && entry.text[lineEnd - 1] == U'\r') {
      --lineEnd;
    }
  }

  return std::u32string_view(entry.text)
      .substr(startOffset, lineEnd - startOffset);
}

} // namespace yuzu::diagnostics

File: crates/yuzu_diagnostics/src/source_map.rs
#[derive(Clone, Copy, PartialEq, PartialOrd)]
pub struct SourceId(usize);

pub struct LineCol {
    pub line: usize,
    pub col: usize,
}

struct Entry {
    name: String,
    text: String,
    line_starts: Vec<usize>,
}

pub struct SourceMap {
    entries: Vec<Entry>,
}

impl Default for SourceMap {
    fn default() -> Self {
        Self::new()
    }
}

impl SourceMap {
    pub fn new() -> Self {
        Self {
            entries: Vec::new(),
        }
    }

    pub fn add(&mut self, name: String, text: String) -> SourceId {
        let line_starts = index_lines(&text);
        self.entries.push(Entry {
            name,
            text,
            line_starts,
        });
        SourceId(self.entries.len())
    }

    pub fn name(&self, source_id: SourceId) -> &str {
        &self.entries[source_id.0 - 1].name
    }

    pub fn text(&self, source_id: SourceId) -> &str {
        &self.entries[source_id.0 - 1].text
    }

    pub fn line_col(&self, source_id: SourceId, offset: usize) -> LineCol {
        let entry = &self.entries[source_id.0 - 1];
        let line_index = entry.line_starts.partition_point(|&start| start <= offset) - 1;

        LineCol {
            line: line_index + 1,
            col: offset - entry.line_starts[line_index] + 1,
        }
    }

    pub fn line_text(&self, source_id: SourceId, line: usize) -> &str {
        let entry = &self.entries[source_id.0 - 1];

        if line == 0 || line > entry.line_starts.len() {
            return "";
        }

        let start = entry.line_starts[line - 1];
        let end = entry
            .line_starts
            .get(line)
            .copied()
            .unwrap_or(entry.text.len());

        let text = &entry.text[start..end];
        let text = text.strip_suffix('\n').unwrap_or(text);
        text.strip_suffix('\r').unwrap_or(text)
    }
}

fn index_lines(text: &str) -> Vec<usize> {
    let mut starts = vec![0];
    starts.extend(text.match_indices('\n').map(|(i, _)| i + 1));
    starts
}
